crates/mirror-mirror/src/try_visit.rs

```diff
@@ -0,0 +1,204 @@
+                ScalarRef::i16(inner) => visitor.try_visit_i16(inner)?,
+                ScalarRef::i32(inner) => visitor.try_visit_i32(inner)?,
+                ScalarRef::i64(inner) => visitor.try_visit_i64(inner)?,
+                ScalarRef::i128(inner) => visitor.try_visit_i128(inner)?,
+                ScalarRef::char(inner) => visitor.try_visit_char(inner)?,
+                ScalarRef::f64(inner) => visitor.try_visit_f64(inner)?,
+                ScalarRef::String(inner) => visitor.try_visit_string(inner)?,
+            }
+        }
+        Type::Struct(struct_ty) => {
+            let struct_ = value.as_struct().unwrap();
+
+            for field_ty in struct_ty.field_types() {
+                let field = struct_.field(field_ty.name()).unwrap();
+                try_visit(visitor, field, field_ty.get_type())?;
+            }
+        }
+        Type::TupleStruct(tuple_struct_ty) => {
+            let tuple_struct = value.as_tuple_struct().unwrap();
+
+            for (idx, field_ty) in tuple_struct_ty.field_types().enumerate() {
+                let field = tuple_struct.field_at(idx).unwrap();
+                try_visit(visitor, field, field_ty.get_type())?;
+            }
+        }
+        Type::Tuple(tuple_ty) => {
+            let tuple = value.as_tuple().unwrap();
+
+            for (idx, field_ty) in tuple_ty.field_types().enumerate() {
+                let field = tuple.field_at(idx).unwrap();
+                try_visit(visitor, field, field_ty.get_type())?;
+            }
+        }
+        Type::Enum(enum_ty) => {
+            let enum_ = value.as_enum().unwrap();
+            let variant_ty = enum_ty.variant(enum_.variant_name()).unwrap();
+
+            for (idx, field_ty) in variant_ty.field_types().enumerate() {
+                let field = match field_ty {
+                    VariantField::Named(named_field_ty) => {
+                        enum_.field(named_field_ty.name()).unwrap()
+                    }
+                    VariantField::Unnamed(_) => enum_.field_at(idx).unwrap(),
+                };
+                try_visit(visitor, field, field_ty.get_type())?;
+            }
+        }
+        Type::List(list_ty) => {
+            let list = value.as_list().unwrap();
+            let element_ty = list_ty.element_type();
+
+            for element in list.iter() {
+                try_visit(visitor, element, element_ty)?;
+            }
+        }
+        Type::Array(array_ty) => {
+            let array = value.as_array().unwrap();
+            let element_ty = array_ty.element_type();
+
+            for element in array.iter() {
+                try_visit(visitor, element, element_ty)?;
+            }
+        }
+        Type::Map(map_ty) => {
+            let map = value.as_map().unwrap();
+            let value_ty = map_ty.value_type();
+
+            for (_, value) in map.iter() {
+                try_visit(visitor, value, value_ty)?;
+            }
```

Comment:
Would we want to allow traversing the map keys too?

Reply:
Yeah we probably should!